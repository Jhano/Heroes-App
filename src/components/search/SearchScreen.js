import React, { useMemo } from 'react';
import  queryString from 'query-string';
import { Form, FormControl, Button, Container, Col,Row, Alert  } from 'react-bootstrap';
import { useLocation } from 'react-router-dom';


import useForm from '../../hooks/useForm';
import HeroCard from '../heroes/HeroCard';
import getHeroesByName from '../selectors/getHeroesByName';


const SearchScreen = ({history}) => {

    //hook para ver la location y para ver las query.
    const location = useLocation();

    //puedo separar todos los querys que mande en el url
    
    const {q = ''} = queryString.parse(location.search); //si q es undefind, se le asigna un string vacio
    
    const [formValues, handleInputChange] =  useForm({
        searchText: q //de esta manera si actualizao la pag el searchText queda con el query de la url
    })

    
    const {searchText} = formValues;

    const heroesFiltered =  useMemo(() => getHeroesByName(q), [q]);

    const handleSubmit = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);
    }

    return ( 
        <Container  style={{marginTop: '20px'}}>
            <Row>
                
                <Col md="4" lg="4">
                    <h1 style={{color: 'white'}}>Search Form</h1>
                    <hr style={{background: 'white'}}/>
                    <Form      
                    style={{justifyContent: 'center'}}
                    onSubmit={handleSubmit}
                    >
                                <Row>
                                    <Col md="12" lg="12">  
                                        <FormControl 
                                            type="text" 
                                            name="searchText"
                                            placeholder="Search"
                                            value={searchText}
                                            style={{ 
                                                    marginBottom:'10px',
                                                }}  
                                            onChange={handleInputChange}           
                                        />
                                    </Col>
                                    <Col md="12" lg="12">
                                        <Button type="submit" 
                                                variant="success"
                                                style={{
                                                        marginBottom:'10px',
                                                        }}
                                                block
                                                
                                        >
                                            Search
                                        </Button>
                                    </Col>
                                    
                                    
                                </Row>
                                
                        </Form>
                    </Col>
                <Col md="8" lg="8">
                <h1 style={{color: 'white'}}>Result</h1>
                <hr style={{background: 'white'}}/>
                        <Row>
                            {
                                (q === '') 
                                    &&
                                    <Col>
                                        <Alert 
                                            variant="info"
                                            style={{textAlign:'center'}}
                                            className="animate__animated animate__fadeIn"
                                        >
                                            Search a hero
                                        </Alert>
                                    </Col>
                                    
                            }

                            {
                                (q !=='' && heroesFiltered.length === 0) 
                                    &&
                                    <Col>
                                        <Alert 
                                            variant="danger"
                                            style={{textAlign:'center'}}
                                            className="animate__animated animate__fadeIn"
                                        >
                                            There is no a hero with {q}
                                        </Alert>
                                    </Col>
                                    
                            }

                            { 
                                heroesFiltered.map(hero => (
                                        <Col 
                                            md="4" 
                                            key={hero.id}
                                            style={{marginBottom:'10px'}}
                                            className="animate__animated animate__fadeIn"   
                                        >
                                        <HeroCard 
                                            
                                            key={hero.id}
                                            {...hero}
                                            className="animate__animated animate__fadeIn"   
                                        />
                                        </Col>
                                ))    
                            }
                         </Row>
                        
                </Col>

            </Row>
            
            
        </Container>
        
     );
}
 
export default SearchScreen;