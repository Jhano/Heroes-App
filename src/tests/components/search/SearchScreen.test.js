import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';

import SearchScreen from '../../../components/search/SearchScreen';



describe('Pruebas en <SearchScreen/>', () => {

    test('debe de mostrarse correctamente con valores por defecto', () => {

        const wrapper = mount (
            //creo la Route porque se necesita estar evaluando en una ruta en particular
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
    })

    test('debe de mostrarse a batman y el input con el valor del queryString', () => {

        const wrapper = mount (
            //creo la Route porque se necesita estar evaluando en una ruta en particular
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        )

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    })

    test('debe de mostrar un error si no se encuentra el hero', () => {

        const busqueda = 'batman123123';

        const wrapper = mount (
            //creo la Route porque se necesita estar evaluando en una ruta en particular
            <MemoryRouter initialEntries={[`/search?q=${busqueda}`]}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        )
        
        expect(wrapper.find('.alert-danger').text().trim()).toBe(`There is no a hero with ${busqueda}`);
        expect(wrapper).toMatchSnapshot();
    })

    test('debe de llamar el push del history', () => {

    

        const historyMock = {
            push: jest.fn(),
        }

        const wrapper = mount (
            //creo la Route porque se necesita estar evaluando en una ruta en particular
            <MemoryRouter initialEntries={[`/search?q=batman`]}>
                <Route path="/search" component={() => <SearchScreen 
                                                        history={historyMock}
                                                        /> }/>
            </MemoryRouter>
        )
        
        wrapper.find('input').simulate('change', {
            target: {
                name:'searchText',
                value:'batman'
            }
        })

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        })

        expect(historyMock.push).toHaveBeenCalledWith(`?q=batman`)
    })
})