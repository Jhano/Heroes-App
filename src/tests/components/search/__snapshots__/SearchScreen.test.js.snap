// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> debe de mostrar un error si no se encuentra el hero 1`] = `
<div
  className="container"
  style={
    Object {
      "marginTop": "20px",
    }
  }
>
  <div
    className="row"
  >
    <div
      className="col-lg-4 col-md-4"
    >
      <h1
        style={
          Object {
            "color": "white",
          }
        }
      >
        Search Form
      </h1>
      <hr
        style={
          Object {
            "background": "white",
          }
        }
      />
      <form
        className=""
        onSubmit={[Function]}
        style={
          Object {
            "justifyContent": "center",
          }
        }
      >
        <div
          className="row"
        >
          <div
            className="col-lg-12 col-md-12"
          >
            <input
              className="form-control"
              name="searchText"
              onChange={[Function]}
              placeholder="Search"
              style={
                Object {
                  "marginBottom": "10px",
                }
              }
              type="text"
              value="batman123123"
            />
          </div>
          <div
            className="col-lg-12 col-md-12"
          >
            <button
              className="btn btn-success btn-block"
              disabled={false}
              style={
                Object {
                  "marginBottom": "10px",
                }
              }
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="col-lg-8 col-md-8"
    >
      <h1
        style={
          Object {
            "color": "white",
          }
        }
      >
        Result
      </h1>
      <hr
        style={
          Object {
            "background": "white",
          }
        }
      />
      <div
        className="row"
      >
        <div
          className="col"
        >
          <div
            className="fade animate__animated animate__fadeIn alert alert-danger show"
            role="alert"
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            There is no a hero with 
            batman123123
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> debe de mostrarse a batman y el input con el valor del queryString 1`] = `
<div
  className="container"
  style={
    Object {
      "marginTop": "20px",
    }
  }
>
  <div
    className="row"
  >
    <div
      className="col-lg-4 col-md-4"
    >
      <h1
        style={
          Object {
            "color": "white",
          }
        }
      >
        Search Form
      </h1>
      <hr
        style={
          Object {
            "background": "white",
          }
        }
      />
      <form
        className=""
        onSubmit={[Function]}
        style={
          Object {
            "justifyContent": "center",
          }
        }
      >
        <div
          className="row"
        >
          <div
            className="col-lg-12 col-md-12"
          >
            <input
              className="form-control"
              name="searchText"
              onChange={[Function]}
              placeholder="Search"
              style={
                Object {
                  "marginBottom": "10px",
                }
              }
              type="text"
              value="batman"
            />
          </div>
          <div
            className="col-lg-12 col-md-12"
          >
            <button
              className="btn btn-success btn-block"
              disabled={false}
              style={
                Object {
                  "marginBottom": "10px",
                }
              }
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="col-lg-8 col-md-8"
    >
      <h1
        style={
          Object {
            "color": "white",
          }
        }
      >
        Result
      </h1>
      <hr
        style={
          Object {
            "background": "white",
          }
        }
      />
      <div
        className="row"
      >
        <div
          className="animate__animated animate__fadeIn col-md-4"
          style={
            Object {
              "marginBottom": "10px",
            }
          }
        >
          <div
            className="card"
          >
            <img
              className="card-img-top"
              src="./assets/heroes/dc-batman.jpg"
              style={
                Object {
                  "height": "16rem",
                }
              }
            />
            <div
              className="card-body"
            >
              <div
                className="card-title h5"
              >
                Batman
              </div>
              <hr />
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <div
                className="card-subtitle h6"
              >
                 
                Detective Comics #27
                 
              </div>
            </div>
            <div
              className="card-footer"
            >
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                <button
                  className="btn btn-info btn-block btn-lg"
                  disabled={false}
                  style={
                    Object {
                      "borderRadius": "2rem",
                    }
                  }
                  type="button"
                >
                   MÃ¡s..
                </button>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> debe de mostrarse correctamente con valores por defecto 1`] = `
<div
  className="container"
  style={
    Object {
      "marginTop": "20px",
    }
  }
>
  <div
    className="row"
  >
    <div
      className="col-lg-4 col-md-4"
    >
      <h1
        style={
          Object {
            "color": "white",
          }
        }
      >
        Search Form
      </h1>
      <hr
        style={
          Object {
            "background": "white",
          }
        }
      />
      <form
        className=""
        onSubmit={[Function]}
        style={
          Object {
            "justifyContent": "center",
          }
        }
      >
        <div
          className="row"
        >
          <div
            className="col-lg-12 col-md-12"
          >
            <input
              className="form-control"
              name="searchText"
              onChange={[Function]}
              placeholder="Search"
              style={
                Object {
                  "marginBottom": "10px",
                }
              }
              type="text"
              value=""
            />
          </div>
          <div
            className="col-lg-12 col-md-12"
          >
            <button
              className="btn btn-success btn-block"
              disabled={false}
              style={
                Object {
                  "marginBottom": "10px",
                }
              }
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="col-lg-8 col-md-8"
    >
      <h1
        style={
          Object {
            "color": "white",
          }
        }
      >
        Result
      </h1>
      <hr
        style={
          Object {
            "background": "white",
          }
        }
      />
      <div
        className="row"
      >
        <div
          className="col"
        >
          <div
            className="fade animate__animated animate__fadeIn alert alert-info show"
            role="alert"
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            Search a hero
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
